### PR TITLE
#13676: improve i0_bw pcc

### PR DESCRIPTION
**Pull request by KalaivaniMCW**
### Ticket
Link to Github Issue #13676

### Problem description
Current implementation of i0_bw uses reciprocal op which has an ongoing issue #14672 

### What's changed
- Reimplemented i0_bw using i1 approach
- implemented ttnn::i1 using ttnn::i0
- Note: yet to add GS and BH implementation - once kernel is finalized
- Updated input range in doc and sweeps

Screenshot 2024-11-21 at 2 41 57 AM

- Profiling :  On main vs On branch

op,count,python min dispatch time (ms),python mean dispatch time(ms),python mean dispatch + sync time (ms),C++ mean dispatch time (ms)
**ttnn.i0_bw,800,0.97,0.997,3.717,0.364** (main )
**ttnn.i0_bw,800,0.241,0.244,0.884,0.086** (branch)

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/11960875200
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp [hunk 1]
```diff
@@ -312,32 +312,43 @@ std::vector<Tensor> ExecuteUnaryBackwardFillZero::invoke(const Tensor& grad, con
...
+    // Tensor value = ttnn::multiply(
+    //     ttnn::multiply(ttnn::i0(input, output_mem_config), ttnn::reciprocal(input, output_mem_config), std::nullopt, output_mem_config),
+    //     0.5,
+    //     std::nullopt,
+    //     output_mem_config);
+    // Tensor result = ttnn::where(
+    //     ttnn::ltz(input, output_mem_config),
+    //     ttnn::multiply(grad,
+    //         ttnn::subtract(ttnn::neg(ttnn::i0(input, output_mem_config), output_mem_config), value, std::nullopt, output_mem_config),
+    //         std::nullopt,
```
Comment (umadevimcw) [inline ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp:337]: Can we remove it?

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward_pybind.hpp [hunk 2]
```diff
@@ -1649,13 +1649,13 @@ void py_module(py::module& module) {
...
         ttnn::i0_bw,
-        R"doc(Performs backward operations for i0 on :attr:`input_tensor` with given :attr:`grad_tensor`.)doc",
+        R"doc(Performs backward operations for i0 on :attr:`input_tensor` with given :attr:`grad_tensor`. Supported input range is (-10, 10))doc",
         R"doc(BFLOAT16, BFLOAT8_B)doc");
 
     detail::bind_unary_backward_op(
         module,
         ttnn::tan_bw,
-        R"doc(Performs backward operations for tan on :attr:`input_tensor` with given :attr:`grad_tensor`. "Supported input range is (-1.45, 1.45)")doc",
+        R"doc(Performs backward operations for tan on :attr:`input_tensor` with given :attr:`grad_tensor`. Supported input range is (-1.45, 1.45))doc",
```
Comment (umadevimcw) [inline ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward_pybind.hpp:1658]: Are these changes necessary in this PR?

File: tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_i1.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+#include "ckernel_defs.h"
+#include "noc_nonblocking_api.h"
+
+using namespace sfpi;
+
+namespace ckernel {
+namespace sfpu {
+
+#define POLYVAL10_DERIVATIVE(coef10, coef9, coef8, coef7, coef6, coef5, coef4, coef3, coef2, coef1, coef0, t4)  \
+    ((coef1 + \
```
Comment (umadevimcw) [inline tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_i1.h:17]: Please write a unit test for forward i1 if possible

**Review by umadevimcw — CHANGES_REQUESTED** [review]

**Review by mcw-anasuya — APPROVED** [review]

File: tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_i1.h (hunk 3 above)
Comment (KalaivaniMCW) [reply tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_i1.h:17]: I haven't enabled pybind for this op yet since it is an intermediate op implemented only for i0_bw

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp (hunk 1 above)
Comment (KalaivaniMCW) [reply ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp:337]: yes will do

**Review by VirdhatchaniKN — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward_pybind.hpp (hunk 2 above)
Comment (KalaivaniMCW) [reply ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward_pybind.hpp:1614]: can be included in doc's PR

**Review by umadevimcw — APPROVED** [review]

**Comment by rdjogoTT** [comment]
@KalaivaniMCW I think the approach of implementing `i1` as a polynomial is correct. I took a quick look at https://github.com/tenstorrent/tt-metal/pull/15325 and found that if I change line 17 in `tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_i1.h` to be:

```
    ((coef0 + (coef1 + (coef2 + (coef3 + (coef4 + (coef5 + (coef6 + (coef7 + (coef8 + (coef9 + coef10 * t2) * t2) * t2) * t2) * t2) * t2) * t2) * t2 ) * t2) * t2) * t2)
```

then the i1 test gets PCC ~= 0.999998 and the i0_bw test gets PCC ~= 0.9998. The change is that you might be missing a `* t2` term in your `POLYVAL10_I1`. Can you double check that this is it? I haven't taken too close of a look so I might be wrong, but I compared the polynomial you have in `i1` to the one you have in `i0` and the `i0` one has this extra term present.

Similarly, I see that in this PR, the `POLYVAL10_DERIVATIVE` you defined does not make use of the `coef0` term. Not sure if this was intentional or not, so please check.

**Comment by KalaivaniMCW** [comment]
@rdjogoTT 
In `POLYVAL10_DERIVATIVE` I pass the `coef0` term just to compare with current implementation. what happens is the `coef0` is * with 0 and so it gets ignored. Let me test `POLYVAL10_I1` once again. Thank you for the feedback!

**Comment by KalaivaniMCW** [comment]
Going with the approach in #15325
